vllm/sequence.py

```diff
@@ -280,11 +289,33 @@ def n_blocks(self) -> int:
 
     @property
     def prompt(self) -> Optional[str]:
-        return self.inputs.get("prompt")
+        if self._prompt is not None:
+            # Reuse precomputed prompt string
+            return self._prompt
+
+        # Select decoder or encoder input prompt str,
+        # as appropriate
+        prompt_key = ("prompt"
```

Comment:
See the new Sequence docstring below, for reference. The intended usage is that `self.from_decoder_prompt` is set at construction-time and never changed for a given sequence; thus, a given sequence can either be a decoder sequence or encoder sequence, mutually-exclusive, based on whether `self.from_decoder_prompt` is set or not. 

Two questions that might logically follow from this:

(1) Why do we need to construct both encoder sequences & decoder sequences?

Here is the code which constructs a `SequenceGroup` inside of `LLMEngine`:

```
# Create the sequence group.
seq_group = SequenceGroup(
    request_id=request_id,
    seqs=[seq],
    arrival_time=arrival_time,
    sampling_params=sampling_params,
    lora_request=lora_request,
    trace_headers=trace_headers,
    prompt_adapter_request=prompt_adapter_request,
    encoder_seq=encoder_seq)
```

You will notice that it takes a `seqs` argument and an `encoder_seq` argument; thus, in order to construct a `SequenceGroup`, I need to be able to first construct both an encoder sequence and a decoder sequence. Since LLMInputs encapsulates both decoder and encoder prompts (in an encoder/decoder scenario), I thus need a way to construct both encoder and decoder sequences from an LLMInputs instance.

(2) In an encoder/decoder scenario, why does a given `Sequence` instance need to hold onto the full `LLMInputs` data structure, given that only one prompt from `LLMInputs` (either encoder or decoder prompt) is utilized by a given sequence?

The reason for holding onto the `LLMInputs` data structure, is that this is already how the current implementation of `Sequence` works: the `Sequence` instance has member which is a reference to the `LLMInputs` instance used to construct it. So all I have done, is to modify the existing implementation so that in an encoder/decoder scenario, it is possible to construct a `Sequence` instance from either the encoder or the decoder prompt.

New Sequence docstring, for reference:

```
"""Stores the data, status, and block information of a sequence.

The sequence is constructed from the LLMInputs instance passed
in through the `inputs` constructor argument.

For encoder/decoder models, LLMInputs encapsulates both a
decoder and encoder prompt, creating an ambiguity about which
prompt to construct the sequence from. The `from_decoder_prompt`
constructor argument signals whether to construct the Sequence
from the LLMInputs decoder prompt, or encoder prompt.

Args:
    seq_id: The ID of the sequence.
    inputs: The inputs of the sequence.
    block_size: The block size of the sequence. Should be the same as the
        block size used by the block manager and cache engine.
    eos_token_id: The end-of-sequence (EOS) token id recognized by this LLM.
    lora_request: LoRA request.
    prompt_adapter_request: Prompt Adapter request.
    from_decoder_prompt: Construct Sequence from LLMInputs decoder prompt
                         (True) or encoder prompt (False.) Must be True
                         for decoder-only model.

"""
```